### PR TITLE
Extend Grid API to allow custom keyboard navigation

### LINKED ISSUE #4054: Enhance keyboard navigation's API
### Description
We need to provide a way to overrite/change the current keyboard behavior in order to have more rich navigation capabilities. For example, to change the default navigation in the grid for `tab key`, to be `column-based` navigation, and not `row-based` or to navigate with keys different than the arrow keys.

#### Provide callbacks for:
- **navigateTo(rowIndex: number, visibleColumnIndex: number, callback({targetType, target: Object }))**
- **getNextCell(currentRowIndex, currentvisibleColumnIndex, callback(IgxColumnComponent))** - returns { rowIndex, visibleColumnIndex } which defines the next cell, that match specific criteria according to the current position
- **getPreviousCell(currentRowIndex, currentvisibleColumnIndex, callback(IgxColumnComponent))** - returns { rowIndex, visibleColumnIndex } which defines the previous cell, that match specific criteria according to the current position
#### Events
Currently we provide the following key interaction events
- [onFocusChange](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/classes/igxgridcomponent.html#onfocuschange) - will be deprecated
- [onKeydownEnterEditMode](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/classes/igxgridcellcomponent.html#onkeydownentereditmode)
- [onKeydownExitEditMode](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/classes/igxgridcellcomponent.html#onkeydownexiteditmode)
- **onGridKeydown** event will be exposed - the event will emit 
IGridKeydownEventArgs {
    targetType: GridKeydownTargetType;
    target: Object;
    event: Event;
    cancel: boolean;
}

### PR DESCRIPTION
**Pull request by ddincheva**
Closes #4054 

### Additional information (check all that apply):
 - [x] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code
 - [ ] This PR includes API docs for newly added methods/properties
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

**Comment by zdrawku** [comment]
After initial testing, everything looks good. I have only two comments below:
- update the Changelog, version 7.2.8 and add a bullet for `igxGrid Keyboard custom navigation` and below list the changes
- rename `perFormAction` private method to `navigate` or something more relevant to the internal navigation with row and col index

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 1]
```diff
@@ -4884,6 +4905,136 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+
+    private getNextDataRowIndex(currentRowIndex): number {
+        if (currentRowIndex === this.verticalScrollContainer.igxForOf.length) {return currentRowIndex; }
+
+        const nextRow = this.verticalScrollContainer.igxForOf.slice(currentRowIndex + 1, this.verticalScrollContainer.igxForOf.length)
+            .find(rec => !rec.expression && !rec.summaries);
+        return nextRow ? this.verticalScrollContainer.igxForOf.indexOf(nextRow) : currentRowIndex;
+    }
+
+    private isValidPosition(rowIndex, colIndex) {
```
Comment (mpavlinov) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Doesn't have return type

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 2]
```diff
@@ -4884,6 +4905,136 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+        } else {
+            if (colIndexes.length === 0 || this.getNextDataRowIndex(currRowIndex) === currRowIndex) {
+                return {rowIndex: this.getNextDataRowIndex(currRowIndex), visibleColumnIndex: curVisibleColIndex};
+            } else {
+                return {rowIndex: this.getNextDataRowIndex(currRowIndex), visibleColumnIndex: colIndexes[0]};
+            }
+        }
+    }
+
+    public getPreviousCell(currRowIndex: number, curVisibleColIndex: number, callback: (IgxColumnComponent) => boolean = null) {
```
Comment (mpavlinov) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Public API. Doesn't have description and code snippet.

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 3]
```diff
@@ -4884,6 +4905,136 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+                    } else {
+                        this.executeCallback(rowIndex, visibleColIndex, cb);
+                    }
+
+                });
+            this.navigation.horizontalScroll(rowIndex).scrollTo(unpinnedIndex);
+        }
+    }
+
+    public getNextCell(currRowIndex: number, curVisibleColIndex: number, callback: (IgxColumnComponent) => boolean = null) {
```
Comment (mpavlinov) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Public API. Doesn't have description and code snippet.

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 4]
```diff
@@ -4884,6 +4905,136 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
         directive.scrollTo(goal);
     }
 
+    public navigateTo(rowIndex: number, visibleColIndex = -1, cb: Function = null) {
```
Comment (mpavlinov) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Public API. Doesn't have description and code snippet.

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 5]
```diff
@@ -4884,6 +4905,136 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+        }
+        const colIndexes = callback ? columns.filter((col) => callback(col)).map(editCol => editCol.visibleIndex).sort((a, b) => a - b) :
+                                columns.map(editCol => editCol.visibleIndex).sort((a, b) => a - b);
+        const nextCellIndex = colIndexes.find(index => index > curVisibleColIndex);
+        if (this.verticalScrollContainer.igxForOf.slice(currRowIndex, currRowIndex + 1)
+                .find(rec => !rec.expression && !rec.summaries) && nextCellIndex !== undefined) {
+            return {rowIndex: currRowIndex, visibleColumnIndex: nextCellIndex};
+        } else {
+            if (colIndexes.length === 0 || this.getNextDataRowIndex(currRowIndex) === currRowIndex) {
+                return {rowIndex: this.getNextDataRowIndex(currRowIndex), visibleColumnIndex: curVisibleColIndex};
```
Comment (mpavlinov) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Please introduce a new type/interface. We don't want anonymous objects. See ["Anonymous objects is a no-no."](https://github.com/IgniteUI/igniteui-angular/pull/4706#discussion_r282767504)

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 6]
```diff
@@ -4884,6 +4905,136 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+        }
+        const colIndexes = callback ? columns.filter((col) => callback(col)).map(editCol => editCol.visibleIndex).sort((a, b) => b - a) :
+                                columns.map(editCol => editCol.visibleIndex).sort((a, b) => b - a);
+        const prevCellIndex = colIndexes.find(index => index < curVisibleColIndex);
+        if (this.verticalScrollContainer.igxForOf.slice(currRowIndex, currRowIndex + 1)
+                .find(rec => !rec.expression && !rec.summaries) && prevCellIndex !== undefined) {
+            return {rowIndex: currRowIndex, visibleColumnIndex: prevCellIndex};
+        } else {
+            if (colIndexes.length === 0 || this.getPrevDataRowIndex(currRowIndex) === currRowIndex) {
+                return {rowIndex: this.getPrevDataRowIndex(currRowIndex), visibleColumnIndex: curVisibleColIndex};
```
Comment (mpavlinov) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Please introduce a new type/interface.

**Pull request merged**